src/Middleware/RequestBodyBufferMiddleware.php

```diff
@@ -29,27 +30,30 @@ public function __construct($sizeLimit = null)
 
     public function __invoke(ServerRequestInterface $request, $stack)
     {
-        $size = $request->getBody()->getSize();
-
-        if ($size === null) {
-            return new Response(411, array('Content-Type' => 'text/plain'), 'No Content-Length header given');
-        }
+        $body = $request->getBody();
 
-        if ($size > $this->sizeLimit) {
+        // request body of known size exceeding limit
+        if ($body->getSize() > $this->sizeLimit) {
             return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
         }
 
-        $body = $request->getBody();
         if (!$body instanceof ReadableStreamInterface) {
```

Comment:
Shouldn't this check be moved up as this type of stream won't be buffered anyway? Why complain about size in this case (shouldn't happen when using the server at all, might as well just throw)?

Reply:
IMO it makes sense to keep this like this in (the rare) case this middleware is used behind another buffering middleware handler, see also https://github.com/reactphp/http/pull/235#issuecomment-336528451 :+1:

Reply:
@clue What I mean is that- if not a `ReadableStreamInterface`, which means we won't buffer it, we're still returning 413 although the middleware is _not_ responsible for handling this request as it is right now.

My suggestion would be to move the `ReadableStreamInterface` to the Beginning of the method?

Reply:
ping @clue could you check my question above? 
What's missing to make make progress with this PR (and 0.8 in general)?

Reply:
Well, technically buffering and limiting sizes are in fact independent responsibilities. For practical reasons however, I think that it makes sense to bring them close together because buffering without applying a limit is actually harmful (potential for trivial DOS otherwise). If we accept the fact that checking size limits here, then I think it makes sense to also support buffered requests here. This means that I think it makes sense to also reject buffered requests that exceed the given size limits.

Accordingly, I think this code LGTM :+1: What do you think about this?

Reply:
> What do you think about this?

@clue sratch head. I'm not getting my point across. We're only buffering and limiting PSR streams. We are however *still* limiting react-only streams. 

This is imho wrong- as the buffering middleware is not responsible for these requests. I'm proposing-  to be explicit- to change order of things like this:

    public function __invoke(ServerRequestInterface $request, $stack)
    {
        $body = $request->getBody();

        if (!$body instanceof ReadableStreamInterface) {
            return $stack($request);
        }

        // request body of known size exceeding limit
        if ($body->getSize() > $this->sizeLimit) {
            return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
        }

Reply:
Yes, it looks like we're talking past one another :man_shrugging: :-)

The `getBody()` method is guaranteed to *always* return an instance of `Psr\Message\StreamInterface`. This very instance *may* also implement the `React\Stream\ReadableStreamInterface` ("streaming state"), but this is not guaranteed (already in "buffered state").

When this middleware handler is used as documented, then the body is always in a "streaming state". If one were to use this middleware handler twice in the same stack (conceived example), then the first will see a body in "streaming state" while the second will see it in "buffered state".

If this conceived(!) second middleware handler uses a smaller size limit (which I would expect to be more of a programming error than intentional), then IMO the second middleware handler should still reject the message in my opinion.

If I understand your above suggestion correctly, then it looks like this would actually allow the body to pass through without applying the second limit at all.

I think it makes sense to add a test for this behavior to ensure that this is properly covered. Can you add a test for this or would you rather want me to add this to this PR? :+1: